common4j/src/main/com/microsoft/identity/common/java/cache/SharedPreferencesAccountCredentialCache.java

```diff
@@ -100,6 +100,14 @@ public SharedPreferencesAccountCredentialCache(
         Logger.verbose(TAG, "Init: " + TAG);
         mSharedPreferencesFileManager = sharedPreferencesFileManager;
         mCacheValueDelegate = accountCacheValueDelegate;
+        if (mSharedPreferencesFileManager.get(SHA1_APPLICATION_IDENTIFIER_ACCESS_TOKEN_CLEARED) == null) {
+            for (final Credential credential : getCredentials()) {
+                if (isSha1ApplicationIdentifierAccessToken(credential)) {
+                    removeCredential(credential);
+                }
+            }
+            mSharedPreferencesFileManager.put(SHA1_APPLICATION_IDENTIFIER_ACCESS_TOKEN_CLEARED, String.valueOf(true));
+        }
```

Comment:
construction of a sharedpreferences kicks off a bakcground thread to load the xml file into memory.  once it's loaded, any get() call is just reading from an in-memory Map<String, String>, but while that async operation is still going on, a call to get() will block waiting for it to finish.  It needs to load the whole file (every account/credential) into memory before any of the values are accessible.

I think as written, this could add some significant delays to construction path, where construction here now requires that async load to complete in order to check the flag.  For OneAuth, since we're moving to the separate implementation, I'm less concerned here. Is this path performance critical for the broker or MSAL? If a token request is going to happen very soon after this, then I don't think it makes a difference, since you'll just pay for the wait microseconds later.  In Oneauth, there may be no token request coming for many minutes after initialization, but we might be unique in that.

Reply:
I'll consult my team and come back to this comment. It doesn't seem to me that there will always be a token request after the cache is initialized, so this is definitely worth considering.

Reply:
Coming back to this... I got some advice from @rpdome and learned two things:

1. It really doesn't matter when the SHA1 app identifier access token is removed, because no method other than mine is going to query for it. And as far as I know, outside of unit tests, there aren't cases where we rely on the size of getCredentials() directly. So, your original suggestion of running the `removeSha1ApplicationIdentifierAccessTokensIfNeeded()` in a separate thread is actually totally fine, and I updated both SPACC and SPACCWMC to now do this. The unit test was updated to just test on `removeSha1ApplicationIdentifierAccessTokensIfNeeded()` now, (since I wasn't able to find a good way to test the method in the constructor due to the threads) but I think they're just as effective.
2. Dome let me know that there's a `OnUpgradeReceiver` class in broker, which, as the name suggests, is meant to run logic upon the app package being updated. So this gives the opportunity to just run `removeSha1ApplicationIdentifierAccessTokensIfNeeded()` once per update, and the calls in the cache constructors can be removed. This all being said, I think I still want to merge in this PR for the sake of getting rid of the SHA1 usage here, and then create a fresh PR with this new implementation in the next week or so (I promise!).